java-checks/src/main/java/org/sonar/java/checks/ArrayCopyLoopCheck.java

```diff
@@ -0,0 +1,309 @@
+
+    public void setCounter(Symbol counter) {
+      this.counter = counter;
+    }
+  }
+
+  @Override
+  public List<Kind> nodesToVisit() {
+    return Arrays.asList(Kind.FOR_STATEMENT, Kind.FOR_EACH_STATEMENT, Kind.WHILE_STATEMENT);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    Context ctx = new Context();
+    boolean issue = false;
+    if (tree.is(Kind.FOR_STATEMENT)) {
+      issue = checkFor((ForStatementTree) tree, ctx);
+    } else if (tree.is(Kind.WHILE_STATEMENT)) {
+      issue = checkWhile((WhileStatementTree) tree, ctx);
+    } else {
+      issue = checkForEach((ForEachStatement) tree);
+    }
+    if (issue) {
+      StatementTree statement = getStatement(tree);
+      reportIssue(statement, "Use \"Arrays.copyOf\", \"Arrays.asList\", \"Collections.add\" or \"System.arraycopy\" instead.");
+    }
+  }
+
+  private static boolean checkFor(ForStatementTree tree, Context ctx) {
+    ListTree<StatementTree> updates = tree.update();
+    if (updates.size() == 1) {
+      StatementTree update = updates.get(0);
+      if (checkUpdate(update, ctx)) {
+        ExpressionTree condition = tree.condition();
+        if (condition != null && checkCondition(condition, ctx)) {
+          StatementTree statement = getStatement(tree);
+          return statement != null && checkStatement(statement, ctx);
+        }
+      }
+    }
+    return false;
+  }
+
+  private static boolean checkWhile(WhileStatementTree tree, Context ctx) {
+    if (tree.statement().is(Kind.BLOCK)) {
+      BlockTree block = (BlockTree) tree.statement();
+      List<StatementTree> body = block.body();
+      if (body.size() == 2) {
+        StatementTree update = body.get(1);
+        if (checkUpdate(update, ctx)) {
+          ExpressionTree condition = tree.condition();
+          if (checkCondition(condition, ctx)) {
+            StatementTree statement = body.get(0);
+            return checkStatement(statement, ctx);
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private static boolean checkForEach(ForEachStatement tree) {
+    ExpressionTree expression = tree.expression();
+    if (isArray(expression)) {
+      StatementTree statement = getStatement(tree);
+      if (statement != null && statement.is(Kind.EXPRESSION_STATEMENT)) {
+        expression = ((ExpressionStatementTree) statement).expression();
+        return isArrayToListCopy(expression, tree.variable());
+      }
+    }
+    return false;
+  }
+
+  private static boolean checkCondition(ExpressionTree tree, Context ctx) {
+    if (tree.is(Kind.LESS_THAN, Kind.LESS_THAN_OR_EQUAL_TO, Kind.GREATER_THAN, Kind.GREATER_THAN_OR_EQUAL_TO, Kind.NOT_EQUAL_TO)) {
+      BinaryExpressionTree comparison = (BinaryExpressionTree) tree;
+      ExpressionTree lhs = comparison.leftOperand();
+      ExpressionTree rhs = comparison.rightOperand();
+      return isInteger(lhs) && isInteger(rhs) && (isIdentifier(lhs) && isCounter(lhs, ctx)) ^ (isIdentifier(rhs) && isCounter(rhs, ctx));
```

Comment:
calling `isInteger` or `isIdentifier` with `isCounter` is redundant - if it's the same symbol, it means it has the same type and that it is identifier.